src/webnn_native/ops/Reduce.cpp

```diff
@@ -40,24 +40,57 @@ namespace webnn_native { namespace op {
         }
     }
 
-    MaybeError Reduce::Validate() {
-        MaybeError maybeError = OperatorBase::Validate();
+    MaybeError Reduce::CalculateShape() {
+        auto inputShape = mInputs[0]->Shape();
+        std::vector<int32_t> reducedShape(inputShape.size()), outputShape;
+        reducedShape = inputShape;
+        auto axes = mAxes;
+        for (size_t i = 0; i < axes.size(); ++i) {
+            if (axes[i] == -1) {
+                axes[i] = inputShape.size() - 1;
+            }
+            reducedShape[axes[i]] = 1;
+        }
+
+        if (!mOptions.keepDimensions) {
+            for (size_t i = 0; i < inputShape.size(); ++i) {
+                // Don't keep reduced dimensions with size of 1.
+                if (!(inputShape[i] != 1 && reducedShape[i] == 1)) {
+                    outputShape.push_back(inputShape[i]);
+                }
+            }
+            if (outputShape.size() == 0) {
+                outputShape = {1};
+            }
+        } else {
+            outputShape = reducedShape;
+        }
+        mOutputs[0]->SetShape(outputShape);
+        return {};
+    }
+
+    MaybeError Reduce::ValidateAndInferOutputInfo() {
+        MaybeError maybeError = OperatorBase::ValidateAndInferOutputInfo();
         if (maybeError.IsError()) {
             return maybeError;
         }
 
+        auto inputShape = mInputs[0]->Shape();
```

Comment:
I propose move it to base if all oprand need to validate the empty.

Reply:
Thanks. But we can't move it to base:
1. Some ops' input shape has been checked whether is 4D/3D...
2. For some op such as Gemm, the third input options.c can have a empty {} shape.